### PR TITLE
Feature/#48/acollow/pm

### PR DESCRIPTION
**Pull request by acollow**
aop.py was modified to add in a class that allows for the calculation of particulate matter at a user provided threshold and relative humidity. Optics files are used to account for the growth of particles with humidity. mietable.py also changed as the ability to grab the bin radius was needed. This addresses and closes issue #48.

The option to run the extinction class at a set relative humidity was also added.

### CODE REVIEW

File: src/pyobs/aop.py [hunk 1]
```diff
@@ -419,6 +477,122 @@ def getAOP(self,what,Species=None,wavelength=None):
...
+            for q in Tracers:
+
+                if self.verbose:
+                    print('   -',q)
+
+                
+                q_conc = (a['AIRDENS'] * a[q]).values * 1000000000 #convert units to kg/m3
+                #rhod is not in all of the standard optics files, and is temporarily added in the yaml portion of the code. 
+                #rhod_ = mie.getAOP('rhod',  bin, rh, wavelength=wavelength).values
+                rhod_ = self.mieTable[s]['rhod'][bin-1]
```
Comment (patricia-nasa) [inline src/pyobs/aop.py:552]: add some comments that explain what each of these variables is, and why you're scaling by 1e6

File: src/pyobs/aop.py [hunk 2]
```diff
@@ -419,6 +477,122 @@ def getAOP(self,what,Species=None,wavelength=None):
...
+                rLow_ = mie.getBinInfo('rLow', bin)*1000000
+                rUp_ = mie.getBinInfo('rUp', bin)*1000000
+                rEff_ = mie.getAOP('rEff', bin, rh, wavelength=wavelength).values*1000000
+                rEff_zero = mie.getBinInfo('rEffDry', bin)*1000000
+		#If necessary, compute the aerodynamic particle radius
+                if aerodynamic:
+                    rLow_ = rLow_ * np.sqrt((rhod_/1000)/self.mieTable[s]['shapefactor'])
```
Comment (patricia-nasa) [inline src/pyobs/aop.py:559]: add a comment here that explains what shapefactor represents

Comment (patricia-nasa) [inline src/pyobs/aop.py:559]: i'm confused about what units you're using here.  why is rhod_ being divided by 1000?

File: src/pyobs/aop.py [hunk 3]
```diff
@@ -419,6 +477,122 @@ def getAOP(self,what,Species=None,wavelength=None):
...
+                    rUp_ = rUp_ * np.sqrt((rhod_/1000)/self.mieTable[s]['shapefactor'])
+                #Find fraction of bin 
+
```
Comment (patricia-nasa) [inline src/pyobs/aop.py:562]: here you can put a check:
```
if rPM is None:
     fPM = 1.0. # adding everything
else:
     do your other checks here
```

File: src/pyobs/aop.py [hunk 4]
```diff
@@ -512,16 +700,20 @@ def CLI_aop():
...
     g = G2GAOP(aer,config=config,mieRootDir=options.rootDir,verbose=options.verbose)
     for w_ in options.wavelengths.split(','):
         w = float(w_)
         if options.aop == 'ext':
-            ds = g.getAOPext(wavelength=w)
+            ds = g.getAOPext(wavelength=w,fixrh=options.fixrh)
         elif options.aop == 'rt':
             ds = g.getAOPrt(wavelength=w,vector=options.vector)
+        elif options.aop == 'pm':
+            ds = g.getPM(wavelength=w,pmsize=options.d_pm,fixrh=options.fixrh,aerodynamic=options.aerodynamic)
```
Comment (patricia-nasa) [inline src/pyobs/aop.py:716]: remove wavelength here.  it's not used.

File: src/pyobs/aop.py [hunk 5]
```diff
@@ -539,21 +731,23 @@ def Test_g2g_aop():
...
+    aer = xr.open_mfdataset(aer_Nv) # still having trouble with parallel
+    print('aer is loaded')
     g = G2GAOP(aer,mieRootDir=data,verbose=True)
-    rts = None # g.getAOPrt(wavelength=550,vector=False)
-    rtv = g.getAOPrt(wavelength=550,vector=True)
-    ext = None # g.getAOPext(wavelength=550)
-
-    return (g, rts, rtv, ext)
+    print('g is done')
+    #rts = None # g.getAOPrt(wavelength=550,vector=False)
```
Comment (patricia-nasa) [inline src/pyobs/aop.py:742]: why did you take out the other examples. you should add yours to the already existing examples.

File: src/pyobs/mietable.py [hunk 6]
```diff
@@ -14,7 +14,7 @@
       rh               (r) RH values   [fraction]
       rLow             (b) Dry upper radius [m]
       rEffDry          (b) Dry Effective radius [m]
-	  rUp              (b) Dry lower radius [m]
+      rUp              (b) Dry lower radius [m]
```
Comment (patricia-nasa) [inline src/pyobs/mietable.py:17]: I think this is a typo.  rLow is the lower radius, rUp is the upper radius

**Review by patricia-nasa — CHANGES_REQUESTED** [review]

**Comment by patricia-nasa** [comment]
can you resolve the conflict before the next PR.

File: src/pyobs/aop.py [hunk 7]
```diff
@@ -419,6 +477,122 @@ def getAOP(self,what,Species=None,wavelength=None):
 
         raise AOPError("not implemented yet")
 
+    def getPM(self,Species=None,wavelength=None,pmsize=None,fixrh=None,aerodynamic=False):
```
[the remark this answers is not in the document]
Comment (acollow) [reply src/pyobs/aop.py:480]: A dummy wavelength is needed to use mie.getAOP to find the effective radius.

[the remark this answers is not in the document]
Comment (patricia-nasa) [reply src/pyobs/aop.py:480]: it shouldn't require it.  if it does, then mietables.py needs to be fixed.

[the remark this answers is not in the document]
Comment (acollow) [reply src/pyobs/aop.py:480]: Fair point, we could take rMass and rEff out of getAOP and put in a new function to get those with just bin and RH as the input.

[the remark this answers is not in the document]
Comment (patricia-nasa) [reply src/pyobs/aop.py:480]: not necessary.  getAOP already doesn't require wavelength.  if for some reason it's not working for the rEff and rMass case then just fix that.

**Review by patricia-nasa — APPROVED** [review]
Looks good.

**Pull request merged**